<a id='top'></a>

# CSCI 3022: Intro to Data Science - Fall 2019 Practicum 2
***

This practicum is due on Canvas by **11:59 PM on Wednesday December 11**. Your solutions to theoretical questions should be done in Markdown/MathJax directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  

**Here are the rules:** 

1. All work, code and analysis, must be your own. 
1. You may use your course notes, posted lecture slides, textbooks, in-class notebooks, and homework solutions as resources.  You may also search online for answers to general knowledge questions like the form of a probability distribution function or how to perform a particular operation in Python/Pandas. 
1. This is meant to be like a coding portion of your midterm exam. So, the instructional team will be much less helpful than we typically are with homework. For example, we will not check answers, help debug your code, and so on.
1. If something is left open-ended, it is because we want to see how you approach the kinds of problems you will encounter in the wild, where it will not always be clear what sort of tests/methods should be applied. Feel free to ask clarifying questions though.
2. You may **NOT** post to message boards or other online resources asking for help.
3. You may **NOT** copy-paste solutions *from anywhere*.
4. You may **NOT** collaborate with classmates or anyone else.
5. In short, **your work must be your own**. It really is that simple.

Violation of the above rules will result in an immediate academic sanction (*at the very least*, you will receive a 0 on this practicum or an F in the course, depending on severity), and a trip to the Honor Code Council.

**By submitting this assignment, you agree to abide by the rules given above.**

***

**Name**: Elijah Berumen 

***


**NOTES**: 

- You may not use late days on the practicums nor can you drop your practicum grades. 
- If you have a question for us, post it as a **PRIVATE** message on Piazza.  If we decide that the question is appropriate for the entire class, then we will add it to a Practicum clarifications thread. 
- Do **NOT** load or use any Python packages that are not available in Anaconda 3.6. 
- Some problems with code may be autograded.  If we provide a function API **do not** change it.  If we do not provide a function API then you're free to structure your code however you like. 
- Submit only this Jupyter notebook to Canvas.  Do not compress it using tar, rar, zip, etc. 
- This should go without saying, but... For any question that asks you to calculate something, you **must show all work to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit.

---
**Shortcuts:**  [Problem 1](#p1) | [Problem 2](#p2) | [Bottom](#bot)

---

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from calendar import month_name, different_locale
from scipy import stats
import statsmodels.api as sm
from scipy.special import binom

<br>

---
<a id='p1'></a>
[Back to top](#top)

### [40 points] Problem 1:  Amazon Forest Fires

A non-profit orgranization is trying to protect the Amazon rain forest and has recruited you as their head data scientist. For your first task, they've given you a dataset with the number of fires in each state in the Amazon region during each month between 1998 and 2017. They would like to have a 95% confidence interval for the true median number of forest fires that occur in each state on a yearly basis. 

In [46]:
#Starter Code
df = pd.read_csv('amazon.csv', thousands='.', decimal ='/', engine='python')
df.info()

df.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6454 entries, 0 to 6453
Data columns (total 5 columns):
year      6453 non-null object
state     6452 non-null object
month     6454 non-null object
number    6448 non-null float64
date      6450 non-null object
dtypes: float64(1), object(4)
memory usage: 252.2+ KB


year state    month  number        date
0  1998  Acre  Janeiro     0.0  1998-01-01
1  1999  Acre  Janeiro     0.0  1999-01-01
2  2000  Acre  Janeiro     0.0  2000-01-01
3  2001  Acre  Janeiro     0.0  2001-01-01
4  2002  Acre  Janeiro     0.0  2002-01-01
5  2003  Acre  Janeiro    10.0  2003-01-01
6  2004  Acre  Janeiro     0.0  2004-01-01
7  2005  Acre  Janeiro    12.0  2005-01-01
8  2006  Acre  Janeiro     4.0  2006-01-01
9  2007  Acre  Janeiro     0.0  2007-01-01

**Part A:**  This dataset isn't paticularly useful in it's current state, so we'll need to clean it up a bit. Some data scientists say that most of their job is to wrangle data, so this will give you a taste of cleaning a real world data set. Perform the following tasks:
1. Drop the 'date' column. The only information this column holds is the year, which we already have in another column. 
2. Drop any rows with null values in any of the remaining columns
3. Print all the unique values of the 'month' column. You'll notice that one is encoded with a differant character encoding then the format that pandas is using.
3. Convert the Portugese month names to English month names. We've included the 'month_name' and the 'different_encoding' modules of the python calendar library in the top cell above, if you would like to use them. There are many ways to accomplish this task, and these modules are not required, but may make things easier. As part of this step, you should make sure that the Portugese month with the encoding problem is translated to the correct English month. 
4. Check the number column for any values that seem impossible. Drop any negative or fractional values, or any values over 50,000. 50,000 is large enough that no Brazilian state would ever have that many forest fires in one month, so we should get rid of anything above 50000. 
5. Since you're new on the job, some of your co-workers may have played a prank on you... Print out all the unique values of the 'year' column and drop any rows with values that don't make sense.
6. Find the total number of rows remaining after you've done all of the above and write it out in a markdown cell. if you have correctly performed all of the tasks above, your dataframe should now have 6438 rows.

**NOTE:** Since some of these tasks are not totally trivial, you may use any resources other than your classmates on this part of this problem. This means you may consult google, stack overflow, the python/pandas documentation, some random book on pandas you might have, etc... But you may not consult your classmates for help. ***CITE ALL RESOURCES USED IN A CODE COMMENT. A URL OR A BOOK TITLE IS SUFFICIENT. ANY CODE OBIVOUSLY COPIED FROM OUTSIDE SOURCES WITH OUT A CITATION WILL EARN YOU NO CREDIT ON THIS PROBLEM. YOU ARE ALLOWED TO USE THESE RESOURCES ONLY ON THIS PART OF THIS PROBLEM!***

In [75]:
#1 Dropping date column
df.drop(['date'], axis=1) 
#2 Drop empty value rows
df.dropna()
#3 Unique Months
months = df.month.unique()
print(months)
#4 Replace month names in English
df['month'] = np.where(df['month'] == "Janeiro", "January", df['month'])
df['month'] = np.where(df['month'] == "Fevereiro", "February", df['month'])
df['month'] = np.where(df['month'] == "Mar�o", "March", df['month'])
df['month'] = np.where(df['month'] == "Abril", "April", df['month'])
df['month'] = np.where(df['month'] == "Maio", "May", df['month'])
df['month'] = np.where(df['month'] == "Junho", "June", df['month'])
df['month'] = np.where(df['month'] == "Julho", "July", df['month'])
df['month'] = np.where(df['month'] == "Agosto", "August", df['month'])
df['month'] = np.where(df['month'] == "Setembro", "September", df['month'])
df['month'] = np.where(df['month'] == "Outubro", "October", df['month'])
df['month'] = np.where(df['month'] == "Novembro", "November", df['month'])
df['month'] = np.where(df['month'] == "Dezembro", "December", df['month'])
print("Row Count: {}".format(df.count()))

#5 Remove unreasonable rows
# Get values that are not possible
indexNames = df[ df['number'] > 50000].index
# Delete these rows from df
df.drop(indexNames , inplace=True)

# Get values that are not possible
indexNames2 = df[ df['number'] < 0].index
# Delete these rows from df
df.drop(indexNames2 , inplace=True)

# Get values that are not possible
#indexNames3 = df[ df['number'] ].index
# Delete these rows from df
#df.drop(indexNames3 , inplace=True)


#6 Print unique values of the year
# To see weird values

indexNames4 = df[ df['year'] == '1000bc'].index
df.drop(indexNames4, inplace=True)

indexNames5 = df[ df['year'] == '-40'].index
df.drop(indexNames5, inplace=True)

indexNames6 = df[ df['year'] == '10bc'].index
df.drop(indexNames6, inplace=True)

#dropna for the nan?
df.dropna()

indexNames8 = df[ df['year'] == "our new data scientist won't notice this"].index
df.drop(indexNames8, inplace=True)

#df.year.unique()
#7 Row count is below

['January' 'February' 'Mar�o' 'April' 'May' 'June' 'July' 'August'
 'September' 'October' 'November' 'December']
Row Count: year      6447
state     6446
month     6448
number    6442
date      6444
dtype: int64


**Part B:** Extract the median number of forest fires per month, yearly, by state. Store these median values in the given python dictionary.

For Example:

If one year of one state had the following numbers of fires:

Jan: 1

Feb: 2

Mar: 3

Apr: 4

May: 5

Jun: 6

Jul: 7

Aug: 8

Sep: 9

Oct: 10

Nov: 11

Dec: 12

Then the median number of forest fires per month would be 6.5 (the average of the two middle elements since this has an even length)

If the state of "test" had 5 years of recorded data, with the following median forest fire values: \[1, 2, 7, 9, 3\],  then python dictionary should look like: 

\{

    "test": [1, 2, 7, 9, 3]
    
\}

Below we've given you one of the states values in a test, so you can ensure you are calculating the medians correctly.

In [76]:
#GIVEN DICTIONARY CODE
median_num_fires_monthly_yearly = {
    'Acre': None,
    'Alagoas': None,
    'Amapa': None,
    'Amazonas': None,
    'Bahia': None,
    'Ceara': None,
    'Distrito Federal': None,
    'Espirito Santo': None,
    'Goias': None,
    'Maranhao': None,
    'Mato Grosso': None,
    'Minas Gerais': None,
    'Para': None,
    'Paraiba': None,
    'Pernambuco': None,
    'Piau': None,
    'Rio': None,
    'Rondonia': None,
    'Roraima': None,
    'Santa Catarina': None,
    'Sao Paulo': None,
    'Sergipe': None,
    'Tocantins': None  
}

In [116]:
# Your code here
medians = df.groupby(['state', 'year'])['number'].median()
#print(medians)
df2 = medians
for state, group in df2.index: # This allows us to iterate over a new dataframe that is groupings of the old one
    newlist = []
    for item in df2.loc[state]: # we use the temp newlist to fill in the entries for each state for each year for each key in dict
        newlist.append(item) # We fill the dict using state as the key
        
    median_num_fires_monthly_yearly[state] = newlist
    #print (df2.loc[state]

#print(median_num_fires_monthly_yearly)




In [119]:
#Given Test
assert median_num_fires_monthly_yearly['Acre'] == \
[1.5, 0.0, 1.0, 0.5, 1.0, 8.0, 7.0, 13.0, 6.0, 4.5, 0.0, 2.5, 6.0, 7.0, 5.0, 13.5, 12.0, 24.0, 33.5, 45.0] \
, "something is wrong here"

In [120]:
#DONT CHANGE THIS. WE USE IT TO MAKE THE OUTPUT LEGIBLE FOR GRADING
import pprint
pp = pprint.PrettyPrinter(indent=1)
pp.pprint(median_num_fires_monthly_yearly)

{'Acre': [1.5,
          0.0,
          1.0,
          0.5,
          1.0,
          8.0,
          7.0,
          13.0,
          6.0,
          4.5,
          0.0,
          2.5,
          6.0,
          7.0,
          5.0,
          13.5,
          12.0,
          24.0,
          33.5,
          45.0],
 'Alagoas': [0.0,
             4.0,
             7.5,
             3.0,
             13.0,
             12.5,
             6.5,
             11.5,
             8.0,
             4.5,
             18.0,
             21.5,
             10.5,
             20.0,
             24.5,
             14.5,
             12.5,
             36.0,
             13.5,
             8.5],
 'Amapa': [0.0,
           1.5,
           1.0,
           0.5,
           5.0,
           8.0,
           7.0,
           3.0,
           4.0,
           0.0,
           2.0,
           4.0,
           2.0,
           4.0,
           13.0,
           2.5,
           7.0,
           9.5,
           12.0,
           3.0

**Part C:** Since we cannot rely on the central limit thereom for the median, we'll bootstrap some samples. Bootstrap 1000 samples for each state. Each bootstrapped sample should have 50 values drawn from the original sample.

Find the median of each bootstrapped sample, and add it to a list. Save the list of median values for the states of **Sao Paulo** and **Goias**. We'll use them later to plot in part D. Then determine the 95% confidence interval of the true median from each list of bootstrapped medians for each state. Add a list of the low and high values of the confidence interval to the given python dictionary below. To help you check your work, the confidence interval of the state of Acre should be from roughly 4 to 7. Your values will change though, because each bootstrapped sample is picked randomly from the original.

For Example:

If the 95% confidence interval on the median is from 6 to 22 for the state of "test", then the dictionary would look like:

{

    "test": [6, 22]
    
}

In [189]:
#GIVEN CODE
median_num_fires_bootstrap = {
    'Acre': None,
    'Alagoas': None,
    'Amapa': None,
    'Amazonas': None,
    'Bahia': None,
    'Ceara': None,
    'Distrito Federal': None,
    'Espirito Santo': None,
    'Goias': None,
    'Maranhao': None,
    'Mato Grosso': None,
    'Minas Gerais': None,
    'Para': None,
    'Paraiba': None,
    'Pernambuco': None,
    'Piau': None,
    'Rio': None,
    'Rondonia': None,
    'Roraima': None,
    'Santa Catarina': None,
    'Sao Paulo': None,
    'Sergipe': None,
    'Tocantins': None  
}

In [192]:
def bootstrap(data = median_num_fires_monthly_yearly, n=1000, func=np.median):
    """
    n bootstraps with func at each sampling
    """
    simulations = []
    sample_size = len(data)
    xbar_init = np.mean(data)
    for c in range(n):
        itersample = np.random.choice(data, size=sample_size, replace=True)
        simulations.append(func(itersample))
    simulations.sort()
    def ci(p):
        """
        2 sided CI
        """
        u_pval = (1+p)/2.
        l_pval = (1-u_pval)
        l_indx = int(np.floor(n*l_pval))
        u_indx = int(np.floor(n*u_pval))
        return(simulations[l_indx],simulations[u_indx])
    for key in median_num_fires_bootstrap:
        median_num_fires_bootstrap[key] = ci(.95)
        
    return simulations

In [193]:
#DONT CHANGE THIS. WE USE IT TO MAKE THE OUTPUT LEGIBLE FOR GRADING
import pprint
pp = pprint.PrettyPrinter(indent=1)
pp.pprint(median_num_fires_bootstrap)

{'Acre': None,
 'Alagoas': None,
 'Amapa': None,
 'Amazonas': None,
 'Bahia': None,
 'Ceara': None,
 'Distrito Federal': None,
 'Espirito Santo': None,
 'Goias': None,
 'Maranhao': None,
 'Mato Grosso': None,
 'Minas Gerais': None,
 'Para': None,
 'Paraiba': None,
 'Pernambuco': None,
 'Piau': None,
 'Rio': None,
 'Rondonia': None,
 'Roraima': None,
 'Santa Catarina': None,
 'Sao Paulo': None,
 'Sergipe': None,
 'Tocantins': None}


**Part D:** Plot a histogram of the frequency of different median values for the two states **Sao Paulo** and **Goias**. Overlay these histograms on the same plot. Include axis labels, a title, a legend, etc. Choose two colors that work well together and provide enough contrast (e.g. No one can see gold overlayed with yellow), and use reasonable values of the **alpha** parameter so you can see both histograms. Plot two vertical lines that represent the outer bounds of the 95% confidence interval on the true median for each state, in the same color as the state. Does the data for the median look normally distributed? Why or why not? Does this validate our decision to not use the central limit theorem and instead bootstrap our median samples? Explain in a markdown cell below.

In [ ]:
#Your code here.

<br>

---
<a id='p2'></a>
[Back to top](#top)

### [40 points] Problem 2:  Sharknado Prediction

Governor Hickenlooper has charged you with the task of assessing the factors associated with sharknado risk in Colorado. As everyone knows, sharknadoes are a leading cause of sharknado-related illness, and you are a world-renowned data/shark scientist.

You decide to use multiple linear regression to understand and predict what factors lead to increased sharknado hazard. Your lead scientist, aptly named Fin, has collected lots of relevant data at a local sharknado hotspot, the Boulder Reservoir[\*](#footnote). The data cover a variety of sharknado-related environmental and other conditions, and you'll find this data in the file `sharknadoes.csv`. 

**Response**: 

- $\texttt{sharknado hazard}$: the hazard of a sharknado, where 1 is very unlikely and 100 is highly likely

**Features**: 

- $\texttt{taunts}$: the number of times over the past year that someone has taunted a shark
- $\texttt{clouds}$: what percentage of the sky was covered by clouds (fraction, 0-1)
- $\texttt{precipitation}$: amount of precipitation in the past 72 hours (inches)
- $\texttt{earthquake}$: the intensity of the most recent earthquake measured in the continental United States
- $\texttt{shark attacks}$: the number of shark attacks within 72 hours prior to the observation
- $\texttt{ice cream sold}$: the number of units of ice cream sold at the beach concession stand 
- $\texttt{misery index}$: an economic indicator for how miserable the average United States citizen is, based on the unemployment rate and the inflation rate. More [here](https://www.stuffyoushouldknow.com/podcasts/whats-the-misery-index.htm) and [here](https://en.wikipedia.org/wiki/Misery_index_(economics)). Higher values correspond to more miserable citizens.
- $\texttt{temperature}$: the outside temperature, measured in degrees Fahrenheit
- $\texttt{humidity}$: relative humidity (percent, 0-100)
- $\texttt{pizzas sold}$: the number of pizzas sold at the beach concession stand in the past year
- $\texttt{pressure}$: local air pressure (millibar) 
- $\texttt{octopuses}$: the number of octupuses in the vicinity on the day of the observation
- $\texttt{Zach's shoe size}$: the size of the shoes Zach was wearing when the observation was made
- $\texttt{Rachel's shoe size}$: the size of the shoes Rachel was wearing when the observation was made

**Part A**: Read the data from `sharknadoes.csv` into a Pandas DataFrame.  Note that since we will be doing a multiple linear regression we will need all of the features. To make sure the data is "clean", drop any row in the DataFrame that is missing data. 

In [9]:
#Your code here.

# Reading data
df2 = pd.read_csv("sharknadoes.csv")
df2 = df2.dropna()
df2.head(10)

clouds  earthquake  pizzas sold  taunts  pressure  shark attacks  \
0     1.00         7.1       5560.0    15.0    847.12            2.0   
1     1.00         7.4       5179.0    20.0    844.34            4.0   
2     1.00         7.0       5227.0     0.0    839.48            9.0   
3     0.13         7.9       5226.0    34.0    851.28            2.0   
4     1.00         7.5       5491.0     6.0    852.67            2.0   
6     0.89         8.1       5646.0    25.0    845.73            3.0   
7     1.00         7.1       5476.0    16.0    851.98            2.0   
9     0.40         6.8       4986.0    17.0    846.42            2.0   
10    1.00         6.1       5359.0    15.0    847.81            2.0   
11    1.00         5.6       5404.0    36.0    847.81            1.0   

    octopuses  precipitation  misery index  ice cream sold  humidity  \
0         7.0       0.824059     12.987180           273.0     86.41   
1         5.0       0.993296     16.765435           184.0     96.67   
2         2.0       1.173342     16.494518           141.0     53.85   
3         6.0       0.919291      8.277176           146.0     88.72   
4         4.0       1.729127      5.904750           178.0     63.08   
6         7.0       1.370010      8.585912           186.0     61.54   
7         4.0       0.900053     16.545501           259.0     87.44   
9         6.0       0.876142      6.067241           105.0     67.18   
10        3.0       0.898160      6.789533           317.0     78.72   
11        4.0       0.909993      7.323839           285.0     79.74   

    temperature  Zachs shoe size  Rachels shoe size  sharknado hazard  
0          78.0             42.0                9.0             40.22  
1          89.0             42.0                9.5             36.42  
2          65.0              9.5                9.0             19.54  
3          36.0              9.5               10.0             85.00  
4          72.0             42.0                9.0             56.34  
6          57.0              9.5                9.0             55.42  
7          79.0              9.5                9.0             52.66  
9          41.0              9.5                9.5             53.68  
10         78.0              9.5                9.0             43.84  
11         81.0             42.0                9.0             44.83

**Part B**: Perform the appropriate statistical test at the $\alpha = 0.025$ significance level to determine if _at least one_ of the features is related to the the response $y$.  Clearly describe your methodology and show all computations in Python. 

\begin{array}{ll}
\text{Hypothesis tests are as follows} \\
H_0: \beta_1 = \beta_2 \ldots = \beta_{14} = 0  \\
H_1: \beta_k \neq 0 \textrm{ for at least one between k = 1 and 14}  \\
\end{array}

In [14]:
#Your code here.
y = df2["sharknado hazard"]
Xfull = df2.loc[:,df2.columns != "sharknado hazard"]
Xfull = sm.add_constant(Xfull)
full_model = sm.OLS(y, Xfull).fit()
full_model.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:       sharknado hazard   R-squared:                       0.978
Model:                            OLS   Adj. R-squared:                  0.972
Method:                 Least Squares   F-statistic:                     179.4
Date:                Tue, 03 Dec 2019   Prob (F-statistic):           9.60e-42
Time:                        00:27:16   Log-Likelihood:                -174.23
No. Observations:                  72   AIC:                             378.5
Df Residuals:                      57   BIC:                             412.6
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const             -2549.8985     67.605    -37.718      0.000   -2685.275   -2414.522
clouds               -1.5106      2.566     -0.589      0.558      -6.650       3.628
earthquake            2.5079      0.467      5.367      0.000       1.572       3.444
pizzas sold          -0.0006      0.002     -0.373      0.711      -0.004       0.003
taunts                0.3117      0.042      7.447      0.000       0.228       0.396
pressure              3.0688      0.079     38.850      0.000       2.911       3.227
shark attacks        -0.1151      0.144     -0.797      0.429      -0.404       0.174
octopuses            -0.0749      0.143     -0.524      0.602      -0.361       0.211
precipitation         1.3982      0.930      1.503      0.138      -0.464       3.261
misery index          0.0273      0.080      0.340      0.735      -0.133       0.188
ice cream sold        0.0096      0.008      1.193      0.238      -0.007       0.026
humidity              0.0188      0.027      0.706      0.483      -0.035       0.072
temperature          -0.4426      0.053     -8.396      0.000      -0.548      -0.337
Zachs shoe size       0.0271      0.023      1.175      0.245      -0.019       0.073
Rachels shoe size     0.2814      1.273      0.221      0.826      -2.268       2.831
==============================================================================
Omnibus:                        0.821   Durbin-Watson:                   2.225
Prob(Omnibus):                  0.663   Jarque-Bera (JB):                0.300
Skew:                           0.005   Prob(JB):                        0.861
Kurtosis:                       3.316   Cond. No.                     1.03e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.03e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Provided the information above, we have an F-Stat value of $179.4$ and a corresponding p-value of $9.60*10^{-42}$

The Null Hypothesis is rejected since the p-value is obviously less than our alpha of .025. This also means that we can conclude that at least one feature is related to the response y.

**Part C**: Write a function `backward_select(df, resp_str, maxsse)` that takes in the DataFrame (`df`), the name of the column corresponding to the response (`resp_str`), and the maximum desired sum of squared errors (`maxsse`), and returns a list of feature names corresponding to the most important features via backward selection.  Use your code to determine the reduced MLR model with the minimal number of features such that the SSE of the reduced model is less than 570. At each stage in backward selection you should remove the feature that has the highest p-value associated with the hypothesis test for the given slope coefficient $\beta_k \neq 0$.

Your code should clearly indicate which feature was removed in each stage, and the SSE associated with the model fit before the feature's removal. _Specifically, please write your code to print the name of the feature that is going to be removed and the SSE before its removal_. Afterward, be sure to report all of the retained features and the SSE of the reduced model.

**Note**: The point of this exercise is to see if you can implement **backward_select** yourself.  You may of course use canned routines like statmodels OLS, but you may not call any Python method that explicitly performs backward selection.

In [48]:
def backward_select(df, resp, maxsse):
    #Your code here.
    # The Response
    y = df[resp]
    # possible features
    remaining = list(df.columns[df.columns != resp])

    # backward select(start with all and drop highest p - value, while filling this list)
    goodFeat = remaining

    sse = 0 # This is the sum of squares error for later
    
    # This is the actual backward selection
    stage = 1
    X = df[remaining]
    X = sm.add_constant(X)
    model = sm.OLS(y, X).fit() 
    sse = np.sum((y-model.predict(X))**2)

    while sse < maxsse:
    
        rem_feat = model.pvalues.idxmax()   # candidate feature to remove is one with highest p-value
        new_remaining_features = remaining.copy()
        new_remaining_features.remove(rem_feat)
        X = df[new_remaining_features]      # create feature set without the feature to remove
        X = sm.add_constant(X)
        model = sm.OLS(y, X).fit()
        sse = np.sum((y-model.predict(X))**2) # SSE_reduced
        
        if sse < maxsse:
            remaining.remove(rem_feat)
            print("Stage # {}, removed {} feature, SSE(Sum of squares error) = {:0.4f}".format(stage, rem_feat, sse))
        stage = stage + 1
        
    print("Features left:",remaining)
    
    return remaining

**Part D**: Write down the final multiple linear regression model, including estimated parameters, obtained by your backward selection process. 

In [62]:
goodFeat = backward_select(df2, resp="sharknado hazard", maxsse=570)

y = df2["sharknado hazard"]
Xred = df2.loc[:, goodFeat]
Xred = sm.add_constant(Xred)
red_model = sm.OLS(y, Xred).fit()
red_model.summary()

Stage # 1, removed Rachels shoe size feature, SSE(Sum of squares error) = 533.4003
Stage # 2, removed misery index feature, SSE(Sum of squares error) = 534.6876
Stage # 3, removed octopuses feature, SSE(Sum of squares error) = 536.5070
Stage # 4, removed clouds feature, SSE(Sum of squares error) = 539.4966
Stage # 5, removed pizzas sold feature, SSE(Sum of squares error) = 543.3572
Stage # 6, removed humidity feature, SSE(Sum of squares error) = 547.4142
Stage # 7, removed shark attacks feature, SSE(Sum of squares error) = 552.1549
Stage # 8, removed Zachs shoe size feature, SSE(Sum of squares error) = 564.1069
Features left: ['earthquake', 'taunts', 'pressure', 'precipitation', 'ice cream sold', 'temperature']


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:       sharknado hazard   R-squared:                       0.977
Model:                            OLS   Adj. R-squared:                  0.974
Method:                 Least Squares   F-statistic:                     450.3
Date:                Wed, 04 Dec 2019   Prob (F-statistic):           6.28e-51
Time:                        22:53:25   Log-Likelihood:                -176.27
No. Observations:                  72   AIC:                             366.5
Df Residuals:                      65   BIC:                             382.5
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const          -2542.0392     59.776    -42.526      0.000   -2661.420   -2422.659
earthquake         2.4012      0.435      5.520      0.000       1.532       3.270
taunts             0.3188      0.037      8.534      0.000       0.244       0.393
pressure           3.0618      0.071     42.836      0.000       2.919       3.205
precipitation      1.2559      0.885      1.419      0.161      -0.512       3.023
ice cream sold     0.0099      0.008      1.311      0.194      -0.005       0.025
temperature       -0.4680      0.033    -14.220      0.000      -0.534      -0.402
==============================================================================
Omnibus:                        1.389   Durbin-Watson:                   2.233
Prob(Omnibus):                  0.499   Jarque-Bera (JB):                0.798
Skew:                           0.210   Prob(JB):                        0.671
Kurtosis:                       3.300   Cond. No.                     1.50e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.5e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

We can see in the coef column of the above where const is -2542.0392 the model is as follows

\begin{align}
\hat{y} =  -2542.04 + 2.40*\text{earthquake}+ 0.32*\text{taunts} + 3.06*\text{pressure} 
+ 1.26*\text{precipitation} + 0.01*\text{ice cream}  - 0.47*\text{temperature} 
\end{align}

**Part E**: Perform the appropriate statistical test at the $\alpha = 0.025$ significance level to determine whether there is a statistically significant difference between the full model with all features and the reduced model obtained by backward selection in **Part D**. You may use output from your model fit above, but all calculations should be set up in Markdown/MathJax.

The Null Hypothesis of all non-included features zbove equalling zero will be checked by the Partial F-test as below:

\begin{align}
F = \dfrac{(SSE_{R} - SSE_{R})/(p - k)}{SSE_{F}/(n - p - 1)} -> F_{p-k, n-p-1}
\end{align}

#### **Part F**: Based on your conclusions in **Part E**, use the _better_ of the two models to predict the sharknado hazard when the following features are observed: 

- $\texttt{taunts}$: 47
- $\texttt{clouds}$: 0.8
- $\texttt{precipitation}$: 1 inch
- $\texttt{earthquake}$: 5
- $\texttt{shark attacks}$: 11
- $\texttt{ice cream sold}$: 120
- $\texttt{misery index}$: 15
- $\texttt{temperature}$: 70 degrees F
- $\texttt{humidity}$: 83
- $\texttt{pizzas sold}$: 5500
- $\texttt{pressure}$: 850 millibar 
- $\texttt{octopuses}$: 6
- $\texttt{Zach's shoe size}$: 9.5
- $\texttt{Rachel's shoe size}$: 9

In [75]:
# Dictionary of key value pairs

feats = {"const" : 1, "taunts": 47, "clouds": 0.8, "precipitation": 1, "earthquake": 5, "shark attacks": 11, "ice cream sold": 120, "misery index": 15, "temperature": 70, \
        "humidity": 83, "pizzas sold": 5500, "pressure": 850, "octopuses": 6, \
         "Zachs size": 9.5, "Rachels shoe size": 9}

y = 0
for f, slope in zip(red_model.params.index, red_model.params):
    y += feats[f] * slope # This is y hat
print("y - hat = {:.3f}".format(y)) # Return y hat 

y - hat = 57.192


$\text{We get a Y - Hat(Predicted Value) of 57.192}$

**Part G:** Consider the model you used in Part E, and consider the fact that you are trying to predict **sharknado hazard**. What is one critical drawback to the MLR model (or any MLR model) for predicting shardnado hazard? What are some modifications that could improve on this issue?

\begin{align}
\text{Since this is multiple linear regression, the sharknado predicted value could potentially be outside of our bounds of 1 to 100 which is the range we want for Sharknado hazard.}
\end{align}